File: sdks/java/src/main/java/dev/atap/sdk/crypto/Base64Url.java
package dev.atap.sdk.crypto;

import java.util.Base64;

/**
 * Base64url encoding/decoding without padding (RFC 4648 Section 5).
 */
public final class Base64Url {

    private Base64Url() {
    }

    /**
     * Base64url encode without padding.
     */
    public static String encode(byte[] data) {
        return Base64.getUrlEncoder().withoutPadding().encodeToString(data);
    }

    /**
     * Base64url decode with padding restoration.
     */
    public static byte[] decode(String s) {
        // Add padding if needed
        int padding = 4 - s.length() % 4;
        if (padding != 4) {
            StringBuilder sb = new StringBuilder(s);
            for (int i = 0; i < padding; i++) {
                sb.append('=');
            }
            s = sb.toString();
        }
        return Base64.getUrlDecoder().decode(s);
    }
}

File: sdks/php/src/Crypto/Base64Url.php
<?php

declare(strict_types=1);

namespace Atap\Sdk\Crypto;

/**
 * Base64url encoding/decoding (RFC 4648 Section 5) without padding.
 */
final class Base64Url
{
    /**
     * Encode data using base64url without padding.
     */
    public static function encode(string $data): string
    {
        return rtrim(strtr(base64_encode($data), '+/', '-_'), '=');
    }

    /**
     * Decode base64url-encoded data, restoring padding.
     */
    public static function decode(string $data): string
    {
        $remainder = strlen($data) % 4;
        if ($remainder !== 0) {
            $data .= str_repeat('=', 4 - $remainder);
        }

        $decoded = base64_decode(strtr($data, '-_', '+/'), true);
        if ($decoded === false) {
            throw new \InvalidArgumentException('Invalid base64url data');
        }

        return $decoded;
    }
}
